# see AvoidanceParameters description in avoidance_module_data.hpp for description.
/**:
  ros__parameters:
    avoidance:
      resample_interval_for_planning: 0.3               # [m] FOR DEVELOPER
      resample_interval_for_output: 4.0                 # [m] FOR DEVELOPER

      # drivable lane setting. this module is able to use not only current lane but also right/left lane
      # if the current lane(=lanelet::Lanelet) and the right/left lane share the boundary(=lanelet::Linestring) in HDMap.
      # "current_lane"           : use only current lane. this module doesn't use adjacent lane to avoid object.
      # "same_direction_lane"    : this module uses same direction lane to avoid object if need.
      # "opposite_direction_lane": this module uses both same direction and opposite direction lane.
      use_lane_type: "opposite_direction_lane"
      # drivable area setting
      use_intersection_areas: true
      use_hatched_road_markings: true
      use_freespace_areas: true

      # avoidance is performed for the object type with true
      target_object:
        car:
          th_moving_speed: 1.0                          # [m/s]
          th_moving_time: 2.0                           # [s]
          longitudinal_margin: 0.0                      # [m]
          lateral_margin:
            soft_margin: 0.7                            # [m]
            hard_margin: 0.2                            # [m]
            hard_margin_for_parked_vehicle: 0.7         # [m]
          max_expand_ratio: 0.0                         # [-] FOR DEVELOPER
          envelope_buffer_margin: 0.1                   # [m] FOR DEVELOPER
          th_error_eclipse_long_radius : 0.6            # [m]
        truck:
          th_moving_speed: 1.0
          th_moving_time: 2.0
          longitudinal_margin: 0.0
          lateral_margin:
            soft_margin: 0.7
            hard_margin: 0.2
            hard_margin_for_parked_vehicle: 0.7
          max_expand_ratio: 0.0
          envelope_buffer_margin: 0.1
          th_error_eclipse_long_radius : 0.6
        bus:
          th_moving_speed: 1.0
          th_moving_time: 2.0
          longitudinal_margin: 0.0
          lateral_margin:
            soft_margin: 0.7
            hard_margin: 0.2
            hard_margin_for_parked_vehicle: 0.7
          max_expand_ratio: 0.0
          envelope_buffer_margin: 0.1
          th_error_eclipse_long_radius : 0.6
        trailer:
          th_moving_speed: 1.0
          th_moving_time: 2.0
          longitudinal_margin: 0.0
          lateral_margin:
            soft_margin: 0.7
            hard_margin: 0.2
            hard_margin_for_parked_vehicle: 0.7
          max_expand_ratio: 0.0
          envelope_buffer_margin: 0.1
          th_error_eclipse_long_radius : 0.6
        unknown:
          th_moving_speed: 0.28
          th_moving_time: 1.0
          longitudinal_margin: 0.0
          lateral_margin:
            soft_margin: 0.7
            hard_margin: -0.2
            hard_margin_for_parked_vehicle: -0.2
          max_expand_ratio: 0.0
          envelope_buffer_margin: 0.1
          th_error_eclipse_long_radius : 0.6
        bicycle:
          th_moving_speed: 0.28
          th_moving_time: 1.0
          longitudinal_margin: 1.0
          lateral_margin:
            soft_margin: 0.7
            hard_margin: 0.5
            hard_margin_for_parked_vehicle: 0.5
          max_expand_ratio: 0.0
          envelope_buffer_margin: 0.5
          th_error_eclipse_long_radius : 0.6
        motorcycle:
          th_moving_speed: 1.0
          th_moving_time: 1.0
          longitudinal_margin: 1.0
          lateral_margin:
            soft_margin: 0.7
            hard_margin: 0.3
            hard_margin_for_parked_vehicle: 0.3
          max_expand_ratio: 0.0
          envelope_buffer_margin: 0.5
          th_error_eclipse_long_radius : 0.6
        pedestrian:
          th_moving_speed: 0.28
          th_moving_time: 1.0
          longitudinal_margin: 1.0
          lateral_margin:
            soft_margin: 0.7
            hard_margin: 0.5
            hard_margin_for_parked_vehicle: 0.5
          max_expand_ratio: 0.0
          envelope_buffer_margin: 0.5
          th_error_eclipse_long_radius : 0.6
        lower_distance_for_polygon_expansion: 30.0      # [m] FOR DEVELOPER
        upper_distance_for_polygon_expansion: 100.0     # [m] FOR DEVELOPER

      # For target object filtering
      target_filtering:
        # avoidance target type
        target_type:
          car: true                                     # [-]
          truck: true                                   # [-]
          bus: true                                     # [-]
          trailer: true                                 # [-]
          unknown: true                                 # [-]
          bicycle: true                                 # [-]
          motorcycle: true                              # [-]
          pedestrian: true                              # [-]
        # detection range
        object_check_goal_distance: 20.0                # [m]
        object_check_return_pose_distance: 20.0         # [m]
        # lost object compensation
        max_compensation_time: 2.0

        # detection area generation parameters
        detection_area:
          static: false                                 # [-]
          min_forward_distance: 50.0                    # [m]
          max_forward_distance: 150.0                   # [m]
          backward_distance: 10.0                       # [m]

        # filtering parking objects
        parked_vehicle:
          th_offset_from_centerline: 1.0                # [m]
          th_shiftable_ratio: 0.8                       # [-]
          min_road_shoulder_width: 0.5                  # [m] FOR DEVELOPER

        # for merging/deviating vehicle
        merging_vehicle:
          th_overhang_distance: 0.0                     # [m]

        # params for avoidance of vehicle type objects that are ambiguous as to whether they are parked.
        avoidance_for_ambiguous_vehicle:
          # policy for ego behavior for ambiguous vehicle.
          # "auto"   : generate candidate path. if RTC is running as AUTO mode, the ego avoids it automatically.
          # "manual" : generate candidate path and wait operator approval even if RTC is running as AUTO mode.
          # "ignore" : never avoid it.
          policy: "manual"                                # [-]
          condition:
            th_stopped_time: 3.0                        # [s]
            th_moving_distance: 1.0                     # [m]
          ignore_area:
            traffic_light:
              front_distance: 20.0                     # [m]
            crosswalk:
              front_distance: 20.0                      # [m]
              behind_distance: 0.0                     # [m]
          wait_and_see:
            target_behaviors: ["MERGING", "DEVIATING"]  # [-]
            th_closest_distance: 10.0                   # [m]

        # params for filtering objects that are in intersection
        intersection:
          yaw_deviation: 0.349                          # [rad] (default 20.0deg)

        freespace:
          condition:
            th_stopped_time: 5.0                        # [-]

      # For safety check
      safety_check:
        # safety check target type
        target_type:
          car: true                                     # [-]
          truck: true                                   # [-]
          bus: true                                     # [-]
          trailer: true                                 # [-]
          unknown: false                                # [-]
          bicycle: true                                 # [-]
          motorcycle: true                              # [-]
          pedestrian: true                              # [-]
        # safety check configuration
        enable: true                                    # [-]
        check_current_lane: false                       # [-]
        check_shift_side_lane: true                     # [-]
        check_other_side_lane: false                    # [-]
        check_unavoidable_object: false                 # [-]
        check_other_object: true                        # [-]
        # collision check parameters
        check_all_predicted_path: false                 # [-]
        safety_check_backward_distance: 100.0           # [m]
        hysteresis_factor_expand_rate: 1.5              # [-]
        hysteresis_factor_safe_count: 3                 # [-]
        collision_check_yaw_diff_threshold: 3.1416      # [rad]
        # predicted path parameters
        min_velocity: 1.38                              # [m/s]
        max_velocity: 50.0                              # [m/s]
        time_resolution: 0.5                            # [s]
        time_horizon_for_front_object: 3.0              # [s]
        time_horizon_for_rear_object: 10.0              # [s]
        delay_until_departure: 0.0                      # [s]
        # rss parameters
        extended_polygon_policy: "along_path"           # [-] select "rectangle" or "along_path"
        expected_front_deceleration: -1.0               # [m/ss]
        expected_rear_deceleration: -1.0                # [m/ss]
        rear_vehicle_reaction_time: 2.0                 # [s]
        rear_vehicle_safety_time_margin: 1.0            # [s]
        lateral_distance_max_threshold: 2.0             # [m]
        longitudinal_distance_min_threshold: 3.0        # [m]
        longitudinal_velocity_delta_time: 0.0           # [s]

      # For avoidance maneuver
      avoidance:
        # avoidance lateral parameters
        lateral:
          th_avoid_execution: 0.09                      # [m] FOR DEVELOPER
          th_small_shift_length: 0.101                  # [m] FOR DEVELOPER
          soft_drivable_bound_margin: 0.3               # [m]
          hard_drivable_bound_margin: 0.3               # [m]
          max_right_shift_length: 5.0                   # [m] FOR DEVELOPER
          max_left_shift_length: 5.0                    # [m] FOR DEVELOPER
          max_deviation_from_lane: 0.2                  # [m]
          # approve the next shift line after reaching this percentage of the current shift line length.
          # this parameter is added to allow waiting for the return of shift approval until the occlusion behind the avoid target is clear.
          # this feature can be disabled by setting this parameter to 0.0.
          ratio_for_return_shift_approval: 0.5          # [-]
        # avoidance distance parameters
        longitudinal:
          min_prepare_time: 1.0                         # [s]
          max_prepare_time: 3.0                         # [s]
          min_prepare_distance: 1.0                     # [m]
          min_slow_down_speed: 1.38                     # [m/s]
          buf_slow_down_speed: 0.56                     # [m/s] FOR DEVELOPER
          nominal_avoidance_speed: 8.33                 # [m/s] FOR DEVELOPER
          consider_front_overhang: true                 # [-]
          consider_rear_overhang: true                  # [-]
        # return dead line
        return_dead_line:
          goal:
            enable: true                                # [-]
            buffer: 3.0                                 # [m]
          traffic_light:
            enable: true                                # [-]
            buffer: 3.0                                 # [m]

      # For cancel maneuver
      cancel:
        enable: true                                    # [-]
        force:
          duration_time: 2.0                            # [s]

      # For yield maneuver
      yield:
        enable: true                                    # [-]
        enable_during_shifting: false                   # [-]

      # For stop maneuver
      stop:
        max_distance: 20.0                              # [m]
        stop_buffer: 1.0                                # [m] FOR DEVELOPER

      policy:
        # policy for rtc request. select "per_shift_line" or "per_avoidance_maneuver".
        # "per_shift_line": request approval for each shift line.
        # "per_avoidance_maneuver": request approval for avoidance maneuver (avoid + return).
        make_approval_request: "per_shift_line"
        # policy for vehicle slow down behavior. select "best_effort" or "reliable".
        # "best_effort": slow down deceleration & jerk are limited by constraints.
        #                but there is a possibility that the vehicle can't stop in front of the vehicle.
        # "reliable": insert stop or slow down point with ignoring decel/jerk constraints.
        #             make it possible to increase chance to avoid but uncomfortable deceleration maybe happen.
        deceleration: "best_effort"                     # [-]
        # policy for avoidance lateral margin. select "best_effort" or "reliable".
        # "best_effort": output avoidance path with shorten lateral margin when there is no enough longitudinal
        #                margin to avoid.
        # "reliable": module output avoidance path with safe (rtc cooperate) state only when the vehicle can avoid
        #             with expected lateral margin.
        lateral_margin: "best_effort"                   # [-]
        # if true, module doesn't wait deceleration and outputs avoidance path by best effort margin.
        use_shorten_margin_immediately: true            # [-]

      # --------------------------------------
      # FOLLOWING PARAMETERS ARE FOR DEVELOPER
      # --------------------------------------

      constraints:
        # lateral constraints
        lateral:
          velocity: [1.39, 4.17, 11.1]                   # [m/s]
          max_accel_values: [0.5, 0.5, 0.5]             # [m/ss]
          min_jerk_values: [0.2, 0.2, 0.2]              # [m/sss]
          max_jerk_values: [1.0, 1.0, 1.0]              # [m/sss]

        # longitudinal constraints
        longitudinal:
          nominal_deceleration: -1.0                    # [m/ss]
          nominal_jerk: 0.5                             # [m/sss]
          max_deceleration: -1.5                        # [m/ss]
          max_jerk: 1.0                                 # [m/sss]
          max_acceleration: 0.5                         # [m/ss]
          min_velocity_to_limit_max_acceleration: 2.78  # [m/ss]

      # path generation method. select "shift_line_base" or "optimization_base" or "both".
      # "shift_line_base"  : Create avoidance path based on shift line.
      #                      User can control avoidance maneuver execution via RTC.
      #                      However, this method doesn't support complex avoidance scenario (e.g. S-shape maneuver).
      # "optimization_base": This module selects avoidance target object
      #                      and bpp module clips drivable area based on avoidance target object polygon shape.
      #                      But this module doesn't modify the path shape.
      #                      On the other hand, autoware_path_optimizer module optimizes path shape instead of this module
      #                      so that the path can be within drivable area. This method is able to deal with complex avoidance scenario.
      #                      However, user can't control avoidance manuever execution.
      # "both"             : Use both method.
      path_generation_method: "shift_line_base"

      shift_line_pipeline:
        trim:
          quantize_size: 0.1
          th_similar_grad_1: 0.1
          th_similar_grad_2: 0.2
          th_similar_grad_3: 0.5

      # for debug
      debug:
        enable_other_objects_marker: true
        enable_other_objects_info: true
        enable_detection_area_marker: false
        enable_drivable_bound_marker: false
        enable_safety_check_marker: false
        enable_shift_line_marker: false
        enable_lane_marker: false
        enable_misc_marker: false
